### PR TITLE
Add STDP synapse unit testing

### PR DESCRIPTION
**Pull request by clinssen**
This adds a unit test for the STDP synapse, by comparing the weight evolution due to a barrage of pre- and postsynaptic spikes between NEST and a Python reference implementation.

Test stimuli:
- [x] randomly generated (Poisson distribution) spike trains
- [x] hard-coded edge cases such as simultaneous pre and post spikes

Neuron models:
- [x] iaf_psc_exp
- [x] iaf_cond_exp
- [ ] iaf_psc_exp_ps

Todo:
- [ ] Unit test code cleanup
- [ ] Add "documentation" example

### CODE REVIEW

**Comment by clinssen** [comment]
``test_stdp_multiplicity`` has been fixed. The condition for test failure was not appropriate and rewritten. Performance is much better now. The following figure shows how, for smaller and smaller timestep, the final obtained weights of a "normal" versus a "precise" single STDP synapse converge.

**Before (current master):**

image

**After (code in this PR):**

image

File: models/stdp_synapse.h [hunk 1]
```diff
@@ -217,7 +217,7 @@ inline void
 stdp_synapse< targetidentifierT >::send( Event& e, thread t, const CommonSynapseProperties& )
 {
   // synapse STDP depressing/facilitation dynamics
-  const double t_spike = e.get_stamp().get_ms();
+  const double t_spike = e.get_stamp().get_ms() - e.get_offset();
```
Comment (heplesser) [inline models/stdp_synapse.h:220]: @clinssen This change makes me wonder if we have the same error in other synapse models as well.

**Comment by clinssen** [comment]
The failure and fix in ``test_visualization.py`` is completely unrelated, but was apparently triggered by matplotlib figures that were created in the new test, and still open when ``test_visualisation.py`` is run afterwards, which then starts drawing into an existing figure.

File: models/stdp_synapse.h (hunk 1 above)
Comment (clinssen) [reply models/stdp_synapse.h:220]: Yes, it does:
https://github.com/nest/nest-simulator/blob/720f4c55fec83b2925792dcd936e2d18eb076938/models/clopath_synapse.h#L215
https://github.com/nest/nest-simulator/blob/720f4c55fec83b2925792dcd936e2d18eb076938/models/ht_synapse.h#L189
https://github.com/nest/nest-simulator/blob/720f4c55fec83b2925792dcd936e2d18eb076938/models/jonke_synapse.h#L300
https://github.com/nest/nest-simulator/blob/720f4c55fec83b2925792dcd936e2d18eb076938/models/quantal_stp_synapse.h#L195
https://github.com/nest/nest-simulator/blob/720f4c55fec83b2925792dcd936e2d18eb076938/models/stdp_dopamine_synapse.h#L529
https://github.com/nest/nest-simulator/blob/720f4c55fec83b2925792dcd936e2d18eb076938/models/stdp_nn_pre_centered_synapse.h#L235
https://github.com/nest/nest-simulator/blob/720f4c55fec83b2925792dcd936e2d18eb076938/models/stdp_nn_restr_synapse.h#L229
https://github.com/nest/nest-simulator/blob/720f4c55fec83b2925792dcd936e2d18eb076938/models/stdp_nn_symm_synapse.h#L231
https://github.com/nest/nest-simulator/blob/720f4c55fec83b2925792dcd936e2d18eb076938/models/stdp_pl_synapse_hom.h#L244
https://github.com/nest/nest-simulator/blob/720f4c55fec83b2925792dcd936e2d18eb076938/models/stdp_synapse_hom.h#L281
https://github.com/nest/nest-simulator/blob/720f4c55fec83b2925792dcd936e2d18eb076938/models/stdp_triplet_synapse.h#L243
https://github.com/nest/nest-simulator/blob/720f4c55fec83b2925792dcd936e2d18eb076938/models/tsodyks2_synapse.h#L209
https://github.com/nest/nest-simulator/blob/720f4c55fec83b2925792dcd936e2d18eb076938/models/tsodyks_synapse_hom.h#L256
https://github.com/nest/nest-simulator/blob/720f4c55fec83b2925792dcd936e2d18eb076938/models/urbanczik_synapse.h#L207
https://github.com/nest/nest-simulator/blob/720f4c55fec83b2925792dcd936e2d18eb076938/models/vogels_sprekeler_synapse.h#L198

Comment (heplesser) [reply models/stdp_synapse.h:220]: @clinssen So this means that until now, we have not taken spike time offsets into account in plasticity mechanisms, doesn't it? It does not have any effect on the actual timing of the spike transmitted, since the offset information is passed on to the receiving neuron in the Event object. Could you create a follow-up issue for this? We should involve @abigailm in the discussion.

File: pynest/nest/tests/test_stdp_multiplicity.py [hunk 2]
```diff
@@ -194,33 +200,51 @@ def run_protocol(self, pre_post_shift):
...
+        post_weights = self.run_protocol(pre_post_shift=pre_post_shift)
         w_plain = np.array(post_weights['parrot'])
         w_precise = np.array(post_weights['parrot_ps'])
 
-        assert all(w_plain == w_plain[0]), 'Plain weights differ'
-        dw = w_precise - w_plain
-        dwrel = dw[1:] / dw[:-1]
-        assert all(np.round(dwrel, decimals=3) ==
-                   0.5), 'Precise weights do not converge.'
+        assert all(w_plain == w_plain[0]), 'Plain weights should be independent of timestep, but differ!'
```
Comment (heplesser) [inline pynest/nest/tests/test_stdp_multiplicity.py:237]: I find the error message confusing: "independent" on the one side and "differ" on the other—requires some mental gymnastics to connect the two.

File: pynest/nest/tests/test_stdp_multiplicity.py [hunk 3]
```diff
@@ -194,33 +200,51 @@ def run_protocol(self, pre_post_shift):
...
+        abs_err = np.abs(w_precise - w_plain)
+        assert abs_err[-1] < max_abs_err
+        assert np.all(np.diff(abs_err) < 1), 'Error should decrease with smaller timestep!'
```
Comment (heplesser) [inline pynest/nest/tests/test_stdp_multiplicity.py:240]: Should this be `< 0` then?

File: pynest/nest/tests/test_stdp_synapse.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+import unittest
+from math import exp
+import numpy as np
+
+try:
+    import matplotlib as mpl
+    import matplotlib.pyplot as plt
+    DEBUG_PLOTS = True
+except Exception:
+    DEBUG_PLOTS = False
```
Comment (heplesser) [inline pynest/nest/tests/test_stdp_synapse.py:32]: I am not sure that producing plots just because matplotlib is available is a good idea. We do not need plots when running CI, only when debugging. Wouldn't it be better to either make this switch manual (by editing the file)?

See also the "plotting possible" test in `test_visualisation`.

File: pynest/nest/tests/test_stdp_synapse.py [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+            "mu_plus": 0.0,
+            "mu_minus": 0.0,
+            "tau_plus": self.tau_pre,
+            "Wmax": 15.0,
+            "weight": self.init_weight
+        }
+        self.neuron_parameters = {
+            "tau_minus": self.tau_post,
+        }
+        self.hardcoded_trains_length = 15
```
Comment (heplesser) [inline pynest/nest/tests/test_stdp_synapse.py:69]: Is this the sum of the pre- and post-lengths? If yes, why not compute it from the arrays. Otherwise, comment.

File: pynest/nest/tests/test_stdp_synapse.py [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+        self.hardcoded_pre_times = np.array([1, 5, 6, 7, 9, 11, 12, 13.])
```
Comment (heplesser) [inline pynest/nest/tests/test_stdp_synapse.py:70]: I am not really happy about having just one entry with a decimal point to make it a float array. I'd either give all a decimal point or add a `dtype=float` explicitly.

File: pynest/nest/tests/test_stdp_synapse.py [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+        nest.SetKernelStatus({'resolution': self.resolution})
+
+        neurons = nest.Create(
+            self.nest_neuron_model,
+            2,
+            params=self.neuron_parameters)
+        presynaptic_neuron = neurons[0]
+        postsynaptic_neuron = neurons[1]
+
+        generators = nest.Create(
```
Comment (heplesser) [inline pynest/nest/tests/test_stdp_synapse.py:114]: You can use tuple unpacking here directly.

File: pynest/nest/tests/test_stdp_synapse.py [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+
+        # While the random sequences, fairly long, would supposedly
+        # reveal small differences in the weight change between NEST
+        # and ours, some low-probability events (say, coinciding
+        # spikes) can well not have occured. To generate and
+        # test every possible combination of pre/post precedence, we
+        # append some hardcoded spike sequences:
+        # pre: 1       5 6 7   9    11 12 13
+        # post:  2 3 4       8 9 10    12
+        (
```
Comment (heplesser) [inline pynest/nest/tests/test_stdp_synapse.py:135]: The python magic beginning here is rather too tough for my little head. Are you somehow shifting spike times along the time axis? Why is that necessary, and couldn't you operate directly on the numpy arrays instead of the acrobatics here?

File: pynest/nest/tests/test_stdp_synapse.py [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+
+    def test_stdp_synapse(self):
+        self.dendritic_delay = float('nan')
+        self.init_params()
+        for self.dendritic_delay in [1., self.resolution]:
+            self.init_params()
+            for self.nest_neuron_model in ["iaf_psc_exp",
+                                           "iaf_cond_exp",
+                                           "iaf_psc_exp_ps"]:
+                fname_snip = "_[nest_neuron_mdl=" + self.nest_neuron_model + "]"
```
Comment (heplesser) [inline pynest/nest/tests/test_stdp_synapse.py:345]: What is this?

**Review by heplesser — CHANGES_REQUESTED** [review]
@clinssen This looks quite good, but some of the code should be simplified or better documented; tests need to be comprehensible ;).

File: pynest/nest/tests/test_stdp_multiplicity.py (hunk 2 above)
Comment (clinssen) [reply pynest/nest/tests/test_stdp_multiplicity.py:237]: Replaced with 'Plain weights should be independent of timestep!'

**Comment by heplesser** [comment]
@clinssen Ping :)

File: pynest/nest/tests/test_stdp_synapse.py (hunk 9 above)
Comment (clinssen) [reply pynest/nest/tests/test_stdp_synapse.py]: A snippet inserted into the filename when saving debug plots to disk.

File: pynest/nest/tests/test_stdp_synapse.py (hunk 4 above)
Comment (clinssen) [reply pynest/nest/tests/test_stdp_synapse.py]: The "plotting possible" test in ``test_visualization`` does essentially the same thing. matplotlib is not installed on CI, as far as I understood, that is by design.

**Comment by clinssen** [comment]
I removed all changes related to support for precise spike timing in synapses, and test coverage for precise neurons. This feature discussion will be continued in #2035.

File: testsuite/pytests/test_stdp_synapse.py [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+        self.dendritic_delay = float('nan')
+        self.init_params()
+        for self.dendritic_delay in [1., self.resolution]:
+            self.init_params()
+            for self.nest_neuron_model in ["iaf_psc_exp", "iaf_cond_exp"]:
+                fname_snip = "_[nest_neuron_mdl=" + self.nest_neuron_model + "]"
+                self.do_nest_simulation_and_compare_to_reproduced_weight(fname_snip=fname_snip)
+
+
+def suite():
```
Comment (heplesser) [inline testsuite/pytests/test_stdp_synapse.py:335]: Since we now use pytest, anything from here on down is no longer needed.

File: testsuite/pytests/test_stdp_nn_synapses.py [hunk 11]
```diff
@@ -66,6 +66,18 @@ def setUp(self):
             "tau_minus": 33.7
         }
 
+        # While the random sequences, fairly long, would supposedly
```
Comment (heplesser) [inline testsuite/pytests/test_stdp_nn_synapses.py:69]: Could you shorten this comment, just something like "Hard-coded spike sequences testing every possible combination of pre/post precedence"? Although I am not 100% sure what is meant by "precedence" here—just "order"?

**Review by heplesser — CHANGES_REQUESTED** [review]
@clinssen Just a few more comments.
